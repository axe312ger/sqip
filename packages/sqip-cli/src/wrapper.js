#!/usr/bin/env node

import sqipCLI from './sqip-cli'
;(async () => {
  await sqipCLI()
})()
