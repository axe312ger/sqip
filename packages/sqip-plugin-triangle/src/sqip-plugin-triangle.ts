import { access, writeFile, unlink, readFile } from 'fs/promises'
import { constants } from 'fs'
import path from 'path'
import os from 'os'
import execa from 'execa'
import Debug from 'debug'

import {
  SqipPlugin,
  // parseColor,
  SqipPluginOptions,
  PluginOptions,
  SqipCliOptionDefinition,
  SqipImageMetadata,
  loadSVG,
  parseColor
} from 'sqip'

interface TrianglePluginOptions extends SqipPluginOptions {
  options: Partial<TriangleOptions>
}

interface TriangleOptions extends PluginOptions {
  bl: number // Blur radius
  nf: number // Noise factor
  bf: number // Blur factor
  ef: number // Edge factor
  pr: number // Point rate
  pth: number // Points threshold
  pts: number // Maximum number of points
  so: number // Sobel filter threshold
  sl: boolean // Use solid stroke color
  wf: number // Wireframe mode
  st: number // Stroke width
  gr: boolean // Output in grayscale mode
  bg: string // Background color (specified as hex value)
}

const debug = Debug('sqip-plugin-triangle')

const VENDOR_DIR = path.resolve(__dirname, '..', 'triangle-binaries')

let triangleExecutable = 'triangle'

export default class TrianglePlugin extends SqipPlugin {
  static get cliOptions(): SqipCliOptionDefinition[] {
    return [
      { name: 'bl', type: Number, description: 'Blur radius', defaultValue: 2 },
      {
        name: 'nf',
        type: Number,
        description: 'Noise factor',
        defaultValue: 0
      },
      { name: 'bf', type: Number, description: 'Blur factor', defaultValue: 1 },
      { name: 'ef', type: Number, description: 'Edge factor', defaultValue: 6 },
      {
        name: 'pr',
        type: Number,
        description: 'Point rate',
        defaultValue: 0.075
      },
      {
        name: 'pth',
        type: Number,
        description: 'Points threshold',
        defaultValue: 10
      },
      {
        name: 'pts',
        type: Number,
        description: 'Maximum number of points',
        defaultValue: 6
      },
      {
        name: 'so',
        type: Number,
        description: 'Sobel filter threshold',
        defaultValue: 10
      },
      {
        name: 'sl',
        type: Boolean,
        description: 'Use solid stroke color',
        defaultValue: false
      },
      {
        name: 'wf',
        type: Number,
        description: 'Wireframe mode',
        defaultValue: 0
      },
      {
        name: 'st',
        type: Number,
        description: 'Stroke width',
        defaultValue: 1
      },
      {
        name: 'gr',
        type: Boolean,
        description: 'Output in grayscale mode',
        defaultValue: false
      },
      {
        name: 'bg',
        type: String,
        description:
          'Background color (specified as hex value or one of the extracted palette colors)',
        defaultValue: 'Muted'
      }
    ]
  }

  constructor(options: TrianglePluginOptions) {
    super(options)
    const { pluginOptions } = options

    this.options = {
      bl: 2,
      nf: 0,
      bf: 1,
      ef: 6,
      pr: 0.075,
      pth: 10,
      pts: 6,
      so: 10,
      sl: false,
      wf: 0,
      st: 1,
      gr: false,
      web: false,
      bg: 'Muted',
      ...pluginOptions
    }
  }

  public options: TriangleOptions

  async apply(
    imageBuffer: Buffer,
    metadata: SqipImageMetadata
  ): Promise<Buffer> {
    if (metadata.type === 'svg') {
      throw new Error(
        'Triangle needs a raster image buffer as input. Check if you run this plugin in the first place.'
      )
    }
    await this.checkForTriangle()

    const { filename, palette } = metadata

    const tmpFile = path.join(os.tmpdir(), `${filename}-${Date.now()}`)
    const tmpSvgFile = path.join(os.tmpdir(), `${filename}-${Date.now()}.svg`)

    await writeFile(tmpFile, imageBuffer)

    const bg = parseColor({ color: this.options.bg, palette }).toLowerCase()

    const triangleArgs = [
      '-in',
      tmpFile,
      '-bl',
      this.options.bl.toString(),
      '-nf',
      this.options.nf.toString(),
      '-bf',
      this.options.bf.toString(),
      '-ef',
      this.options.ef.toString(),
      '-pr',
      this.options.pr.toString(),
      '-pth',
      this.options.pth.toString(),
      '-pts',
      this.options.pts.toString(),
      '-so',
      this.options.so.toString(),
      this.options.sl ? '-sl' : null,
      '-wf',
      this.options.wf.toString(),
      '-st',
      this.options.st.toString(),
      this.options.gr ? '-gr' : null,
      this.options.web ? '-web' : null,
      '-bg',
      bg,
      '-out',
      tmpSvgFile
    ].filter((v) => v !== null) as string[]

    await execa(triangleExecutable, triangleArgs)

    const result = await readFile(tmpSvgFile)

    await unlink(tmpFile)
    await unlink(tmpSvgFile)

    const $ = loadSVG(result.toString())
    const $svg = $('svg')

    const $bgRect = $(`<rect width="100%" height="100%" fill="${bg}"/>`)

    $svg.prepend($bgRect)

    metadata.type = 'svg'
    metadata.mimeType = 'image/svg'

    return Buffer.from($.html())
  }

  // Sanity check: use the exit state of 'type' to check for Triangle availability
  async checkForTriangle(): Promise<undefined> {
    const platform = os.platform()
    const trianglePath = path.join(
      VENDOR_DIR,
      `triangle-${platform}-${os.arch()}${platform === 'win32' ? '.exe' : ''}`
    )

    try {
      await access(trianglePath, constants.X_OK)
      debug(`Found triangle binary at ${trianglePath}`)
      triangleExecutable = trianglePath
      return
    } catch (e) {
      // noop
    }

    // Test if Triangle is available as global executable
    try {
      if (platform === 'win32') {
        await execa('where', ['triangle'])
      } else {
        await execa('type', ['triangle'])
      }
    } catch (e) {
      throw new Error(
        'Please ensure that triangle (https://github.com/esimov/triangle, written in Golang) is installed and globally available.'
      )
    }

    debug(`Found globally available triangle binary`)
    triangleExecutable = 'triangle'
  }
}
