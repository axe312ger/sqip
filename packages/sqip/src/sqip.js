// #############################################################################
// #
// # "SQIP" (pronounced \skwÉªb\ like the non-magical folk of magical descent)
// # is a SVG-based LQIP technique - https://github.com/technopagan/sqip
// #
// # Installation as CLI:
// # npm install -g sqip
// #
// # Installation as lib:
// # npm install sqip
// #
// # Requirements:
// # * Node.js >= v6 (https://nodejs.org/en/)
// #
// #############################################################################

import path from 'path'

import Debug from 'debug'
import fs from 'fs-extra'

import { getDimensions, printFinalResult } from './helpers'

const debug = Debug('sqip')

export default async function sqip(options) {
  // Build configuration based on passed options and default options
  const defaultOptions = {
    plugins: [
      { name: 'primitive', options: { numberOfPrimitives: 8, mode: 0 } },
      'blur',
      'svgo',
      'data-uri'
    ],
    shouldThrow: true // @todo do we really need this?
  }
  const config = Object.assign({}, defaultOptions, options)

  // Validate configuration
  if (!config.input) {
    throw new Error(
      'Please provide an input image, e.g. sqip({ input: "input.jpg" })'
    )
  }

  const inputPath = path.resolve(process.cwd(), config.input)

  try {
    await fs.access(inputPath, fs.constants.R_OK)
  } catch (err) {
    throw new Error(`Unable to read input file: ${inputPath}`)
  }

  const dimensions = getDimensions(inputPath)

  // Resolver
  const plugins = await Promise.all(
    config.plugins.map(async plugin => {
      if (typeof plugin === 'string') {
        plugin = { name: plugin }
      }
      const { name } = plugin

      if (!name) {
        throw new Error(
          `Unable to read plugin name from:\n${JSON.stringify(plugin, null, 2)}`
        )
      }
      const moduleName =
        name.indexOf('sqip-plugin-') !== -1 ? name : `sqip-plugin-${name}`
      try {
        debug(`Loading ${moduleName}`)
        const Plugin = await import(moduleName)

        return { ...plugin, Plugin }
      } catch (err) {
        if (err.code === 'MODULE_NOT_FOUND') {
          throw new Error(
            `Unable to load plugin "${moduleName}". Try installing it via:\n\n npm install ${moduleName}`
          )
        }
        throw err
      }
    })
  )

  let svg = config.filename
  for (const { name, options: pluginOptions, Plugin } of plugins) {
    try {
      debug(`Construct ${name}`)
      const plugin = new Plugin({ dimensions, ...options, ...pluginOptions })
      debug(`Apply ${name}`)
      svg = await plugin.apply(svg, { dimensions })
    } catch (err) {
      if (config.shouldThrow) {
        throw err
      }
      console.error(err)
      process.exit(1)
    }
  }

  debug(`Finished`)

  // Write to disk or output result
  if (config.output) {
    const outputPath = path.resolve(config.output)
    // @todo remove all sync calls
    fs.writeFileSync(outputPath, svg)
  } else {
    printFinalResult(dimensions, inputPath, svg)
  }
  return { svg, dimensions }
}
