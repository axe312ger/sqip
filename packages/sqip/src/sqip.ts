import path from 'path'

import Debug from 'debug'
import fs from 'fs-extra'
import imageSize from 'probe-image-size'
import Vibrant from 'node-vibrant'
import sharp from 'sharp'
import termimg from 'term-img'
import Table from 'cli-table3'
import chalk from 'chalk'

import { Palette } from '@vibrant/color'
import { OptionDefinition } from 'command-line-args'

import { locateFiles } from './helpers'

export { loadSVG, parseColor } from './helpers'

const debug = Debug('sqip')

const mainKeys = ['originalWidth', 'originalHeight', 'width', 'height', 'type']

const PALETTE_KEYS: (keyof Palette)[] = [
  'Vibrant',
  'DarkVibrant',
  'LightVibrant',
  'Muted',
  'DarkMuted',
  'LightMuted'
]

interface SqipPluginInterface {
  metadata: SqipImageMetadata
  sqipConfig: SqipConfig
  apply(imageBuffer: Buffer): Promise<Buffer> | Buffer
}

export interface SqipResult {
  content: Buffer
  metadata: SqipImageMetadata
}

export interface SqipPluginOptions {
  pluginOptions: PluginOptions
  options: PluginOptions
  metadata: SqipImageMetadata
  sqipConfig: SqipConfig
}
export interface SqipCliOptionDefinition extends OptionDefinition {
  description?: string
  required?: boolean
  default?: boolean
}

export interface PluginOptions {
  [key: string]: unknown
}

interface PluginResolver {
  name: string
  options?: PluginOptions
}

interface SqipConfig {
  input: string | Buffer
  outputFileName?: string
  output: string
  silent?: boolean
  parseableOutput?: boolean
  plugins: PluginType[]
  width?: number
}

interface ProcessFileOptions {
  buffer: Buffer
  outputFileName: string
  config: SqipConfig
}

interface SqipImageMetadata {
  originalWidth: number
  originalHeight: number
  palette: Palette
  height: number
  width: number
  type: string
  [key: string]: unknown
}

type PluginType = PluginResolver | string

export class SqipPlugin implements SqipPluginInterface {
  public metadata: SqipImageMetadata
  public sqipConfig: SqipConfig
  public options: PluginOptions
  static cliOptions: SqipCliOptionDefinition[]

  constructor(options: SqipPluginOptions) {
    const { sqipConfig, metadata } = options
    this.sqipConfig = sqipConfig || {}
    this.metadata = metadata || {}
    this.options = {}
  }
  apply(imageBuffer: Buffer): Promise<Buffer> | Buffer {
    throw new Error('Not implemented')
    return imageBuffer
  }
}

// Resolves plugins based on a given config
// Array of plugin names or config objects, even mixed.
export async function resolvePlugins(
  plugins: PluginType[]
): Promise<(PluginResolver & { Plugin: typeof SqipPlugin })[]> {
  return Promise.all(
    plugins.map(async (plugin) => {
      if (typeof plugin === 'string') {
        plugin = { name: plugin }
      }
      const { name } = plugin

      if (!name) {
        throw new Error(
          `Unable to read plugin name from:\n${JSON.stringify(plugin, null, 2)}`
        )
      }
      const moduleName =
        name.indexOf('sqip-plugin-') !== -1 ? name : `sqip-plugin-${name}`
      try {
        debug(`Loading ${moduleName}`)
        const Plugin = await import(moduleName)

        return { ...plugin, Plugin: Plugin.default }
      } catch (err) {
        if (err.code === 'MODULE_NOT_FOUND') {
          throw new Error(
            `Unable to load plugin "${moduleName}". Try installing it via:\n\n npm install ${moduleName}`
          )
        }
        throw err
      }
    })
  )
}

async function processFile({
  buffer,
  outputFileName,
  config
}: ProcessFileOptions) {
  const { output, silent, parseableOutput } = config
  const result = await processImage({ buffer, config })
  const { content, metadata } = result
  let outputPath

  debug(`Processed ${outputFileName}`)

  // Write result svg if desired
  if (output) {
    try {
      // Test if output path already exists
      const stats = await fs.stat(output)

      // Throw if it is a file and already exists
      if (!stats.isDirectory()) {
        throw new Error(
          `File ${output} already exists. Overwriting is not yet supported.`
        )
      }
      outputPath = path.resolve(output, `${outputFileName}.svg`)
    } catch (err) {
      // Output directory or file does not exist. We will create it later on.
      outputPath = output
    }

    debug(`Writing ${outputPath}`)
    await fs.writeFile(outputPath, content)
  }

  // Gather CLI output information
  if (!silent) {
    if (outputPath) {
      console.log(`Stored at: ${outputPath}`)
    }

    // Generate preview
    if (!parseableOutput) {
      // Convert to png for image preview
      const preview = await sharp(Buffer.from(content)).png().toBuffer()

      try {
        termimg(preview, {
          fallback: () => {
            // SVG results can still be outputted as string
            if (metadata.type === 'svg') {
              console.log(content.toString())
              return
            }

            // No fallback preview solution yet for non-svg files.
            console.log(
              `Unable to render a preview for ${metadata.type} files on this machine. Try using https://iterm2.com/`
            )
          }
        })
      } catch (err) {
        if (err.name !== 'UnsupportedTerminalError') {
          throw err
        }
      }
    }

    // Metadata
    const tableConfig = parseableOutput
      ? {
          chars: {
            top: '',
            'top-mid': '',
            'top-left': '',
            'top-right': '',
            bottom: '',
            'bottom-mid': '',
            'bottom-left': '',
            'bottom-right': '',
            left: '',
            'left-mid': '',
            mid: '',
            'mid-mid': '',
            right: '',
            'right-mid': '',
            middle: ' '
          },
          style: { 'padding-left': 0, 'padding-right': 0 }
        }
      : undefined

    // Figure out which metadata keys to show
    // @todo why is this unused?
    // const allKeys = [...mainKeys, 'palette']

    const mainTable = new Table(tableConfig)
    mainTable.push(mainKeys)
    mainTable.push(
      mainKeys.map((key) => String(metadata[key]) || 'can not display')
    )
    console.log(mainTable.toString())

    // Show color palette
    const paletteTable = new Table(tableConfig)
    paletteTable.push(PALETTE_KEYS)
    paletteTable.push(
      PALETTE_KEYS.map((key) => metadata.palette[key]?.hex)
        .filter<string>((hex): hex is string => typeof hex === 'string')
        .map((hex) => chalk.hex(hex)(hex))
    )
    console.log(paletteTable.toString())

    Object.keys(metadata)
      .filter((key) => ![...mainKeys, 'palette'].includes(key))
      .forEach((key) => {
        console.log(chalk.bold(`${key}:`))
        console.log(metadata[key])
      })
  }

  return result
}

interface ProcessImageOptions {
  buffer: Buffer
  config: SqipConfig
}

async function processImage({
  buffer,
  config
}: ProcessImageOptions): Promise<SqipResult> {
  const originalSizes = imageSize.sync(buffer)
  const vibrant = Vibrant.from(buffer)
  const palette = await vibrant.quality(0).getPalette()

  if (!originalSizes) {
    throw new Error('Unable to get image size')
  }

  const metadata: SqipImageMetadata = {
    originalWidth: originalSizes.width,
    originalHeight: originalSizes.height,
    palette,
    type: 'unknown',
    width: 0,
    height: 0
  }

  // Load plugins
  const plugins = await resolvePlugins(config.plugins)

  // Interate through plugins and apply them to last returned image
  if (config.width && config.width > 0) {
    // Resize to desired output width
    buffer = await sharp(buffer).resize(config.width).toBuffer()

    const resizedMetadata = await sharp(buffer).metadata()
    metadata.width = resizedMetadata.width || 0
    metadata.height = resizedMetadata.height || 0
  } else {
    // Fall back to original size, keep image as is
    metadata.width = originalSizes.width
    metadata.height = originalSizes.height
  }

  for (const { name, options: pluginOptions, Plugin } of plugins) {
    debug(`Construct ${name}`)
    const plugin = new Plugin({
      sqipConfig: config,
      pluginOptions: pluginOptions || {},
      metadata,
      options: {}
    })
    debug(`Apply ${name}`)
    buffer = await plugin.apply(buffer)
  }

  return { content: buffer, metadata }
}

export default async function sqip(
  options: SqipConfig
): Promise<SqipResult | SqipResult[]> {
  // Build configuration based on passed options and default options
  const defaultOptions = {
    plugins: [
      { name: 'primitive', options: { numberOfPrimitives: 8, mode: 0 } },
      'blur',
      'svgo',
      'data-uri'
    ],
    width: 300,
    parseableOutput: false,
    silent: true
  }

  const config = Object.assign({}, defaultOptions, options)

  const { input, outputFileName, parseableOutput, silent } = config

  if (parseableOutput) {
    chalk.level = 0
  }

  // Validate configuration
  if (!input) {
    throw new Error(
      'Please provide an input image, e.g. sqip({ input: "input.jpg" })'
    )
  }

  // If input is a Buffer
  if (Buffer.isBuffer(input)) {
    if (!outputFileName) {
      throw new Error(
        `${outputFileName} is required when passing image as buffer`
      )
    }
    return processFile({
      buffer: input,
      outputFileName,
      config
    })
  }

  const files = await locateFiles(input)
  debug('Found files:')
  debug(files)

  // Test if all files are accessable
  for (const file of files) {
    try {
      debug('check file ' + file)
      await fs.access(file, fs.constants.R_OK)
    } catch (err) {
      throw new Error(`Unable to read file ${file}`)
    }
  }

  // Iterate over all files
  const results = []
  for (const filePath of files) {
    // Apply plugins to files
    if (!silent) {
      console.log(`Processing: ${filePath}`)
    } else {
      debug(`Processing ${filePath}`)
    }
    const buffer = await fs.readFile(filePath)
    const result = await processFile({
      buffer,
      outputFileName: outputFileName || path.parse(filePath).name,
      config
    })

    results.push(result)
  }

  debug(`Finished`)

  // Return as array when input was array or results is only one file
  if (Array.isArray(input) || results.length === 0) {
    return results
  }
  return results[0]
}

export * from './helpers'
