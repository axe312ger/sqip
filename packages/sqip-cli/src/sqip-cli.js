import Debug from 'debug'
import commandLineArgs from 'command-line-args'
import commandLineUsage from 'command-line-usage'

import sqip, { resolvePlugins } from 'sqip'

const { version } = require('../package.json')

const debug = Debug('sqip-cli')

let optionList = [
  {
    name: 'help',
    alias: 'h',
    description: 'Show help'
  },
  {
    name: 'version',
    description: 'Show version number'
  },
  {
    name: 'plugins',
    alias: 'p',
    type: String,
    multiple: true,
    description: 'One or more plugins. E.g. "-p primitive blur"'
  },
  {
    name: 'input',
    alias: 'i',
    type: String,
    required: true
  },
  {
    name: 'output',
    alias: 'o',
    type: String,
    description:
      'Save the resulting SVG to a file. The svg result will be returned by default.'
  }
]

function showHelp({ optionList }) {
  const sections = [
    {
      header: 'sqip CLI',
      content:
        'Usage: sqip --input [path]\n\n"SQIP" (pronounced \\skwÉªb\\ like the non-magical folk of magical descent) is a SVG-based LQIP technique - https://github.com/technopagan/sqip'
    },
    {
      header: 'Options',
      optionList
    },
    {
      header: 'Examples',
      content: `Output input.jpg image as SQIP
$ sqip --input /path/to/input.jpg

Save input.jpg as result.svg with 25 shapes and no blur
$ sqip -i input.jpg -n 25 -b 0 -o result.svg`
    }
  ]
  const usage = commandLineUsage(sections)
  console.log(usage)
}

const args = commandLineArgs(optionList, { partial: true })

if ('version' in args) {
  console.log(version)
  process.exit(0)
}

let { plugins } = args

// Default cli plugins
if (!plugins) {
  plugins = ['primitive', 'blur', 'svgo'].filter(Boolean)
}

;(async () => {
  debug(`Found plugins:\n`, plugins)
  const resolvedPlugins = await resolvePlugins(plugins)

  // Add new cli options based on enabled plugins
  const pluginOptions = resolvedPlugins.reduce((definitions, plugin) => {
    const {
      name,
      Plugin: { cliOptions }
    } = plugin
    if (cliOptions) {
      return [
        ...definitions,
        ...cliOptions.map(option => ({
          ...option,
          name: `${name}-${option.name}`
        }))
      ]
    }
    return definitions
  }, [])

  optionList = [...optionList, ...pluginOptions]

  debug(`Generated CLI options:\n`, optionList)

  const args = commandLineArgs(optionList, { partial: true })

  debug(`Parsed CLI args:\n`, args)

  if ('help' in args) {
    showHelp({ optionList })
    process.exit(0)
  }

  const missing = optionList
    .filter(({ required }) => required)
    .filter(({ name }) => !args[name])
    .map(({ name }) => name)

  if (missing.length) {
    showHelp({ optionList })
    console.info(
      `\nPlease provide the following arguments: ${missing.join(', ')}`
    )
    process.exit(1)
  }

  const { input, output } = args

  // Build list of plugins with options based on passed arguments
  const pluginsOptions = resolvedPlugins.map(({ name }) => {
    const options = Object.keys(args)
      .filter(args => args.indexOf(`${name}-`) === 0)
      .reduce((optionMap, argName) => {
        const optionName = argName.substr(`${name}-`.length)
        return { ...optionMap, [optionName]: args[argName] }
      }, {})
    return { name, options }
  })

  const options = {
    input,
    output,
    plugins: pluginsOptions
  }

  try {
    await sqip(options)
  } catch (err) {
    console.log(err)
    process.exit(1)
  }
})()
