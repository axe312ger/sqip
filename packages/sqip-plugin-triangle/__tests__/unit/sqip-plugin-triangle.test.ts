import execa, { ExecaChildProcess } from 'execa'
import fs from 'fs/promises'
import os from 'os'
import { Swatch } from '@vibrant/color'

import TrianglePlugin from '../../src/sqip-plugin-triangle'
import { SqipImageMetadata } from 'sqip/src/sqip'

jest.mock('execa')
jest.mock('fs/promises')
jest.mock('os', () => ({
  ...(jest.requireActual('os') as typeof os),
  platform: jest.fn(() => 'unknownOS'),
  arch: jest.fn(() => 'nonExistingArch'),
  cpus: () => [1]
}))

const mockedExeca = execa as jest.MockedFunction<typeof execa>
mockedExeca.mockImplementation(() => {
  const result = {
    stdout:
      '<svg viewBox="0 0 1024 768"><rect fill="#bada5500"/><g></g></rect></svg>'
  } as unknown as ExecaChildProcess<Buffer>
  return result
})

const mockedFsAccess = fs.access as jest.MockedFunction<typeof fs.access>

mockedFsAccess.mockImplementation(async () =>
  Promise.reject(new Error('Mocked: Binary not available'))
)
const mockedOsArch = os.arch as jest.MockedFunction<typeof os.arch>
const mockedOsPlatform = os.platform as jest.MockedFunction<typeof os.platform>

const proccessExitSpy = jest.spyOn(process, 'exit').mockImplementation()

const mockedMetadata: SqipImageMetadata = {
  filename: 'mocked',
  mimeType: 'image/mocked',
  width: 1024,
  height: 640,
  type: 'pixel',
  originalHeight: 1024,
  originalWidth: 640,
  palette: {
    DarkMuted: new Swatch([4, 2, 0], 420),
    DarkVibrant: new Swatch([4, 2, 1], 421),
    LightMuted: new Swatch([4, 2, 2], 422),
    LightVibrant: new Swatch([4, 2, 3], 423),
    Muted: new Swatch([4, 2, 4], 424),
    Vibrant: new Swatch([4, 2, 5], 425)
  }
}
const mockedConfig = {
  input: 'mocked',
  output: 'mocked',
  plugins: ['triangle']
}

describe('checkForTriangle', () => {
  const trianglePlugin = new TrianglePlugin({
    pluginOptions: {},
    options: {},
    sqipConfig: mockedConfig
  })

  afterEach(() => {
    mockedExeca.mockClear()
    mockedFsAccess.mockClear()
    mockedOsArch.mockClear()
    mockedOsPlatform.mockClear()
    proccessExitSpy.mockClear()
  })

  afterAll(() => {
    proccessExitSpy.mockReset()
  })

  test('bundled executable exists', async () => {
    mockedFsAccess.mockImplementationOnce(async () => Promise.resolve())

    await trianglePlugin.checkForTriangle()

    expect(global.process.exit).not.toHaveBeenCalled()
    expect(mockedExeca).not.toHaveBeenCalled()
  })

  test('uses where for windows, type for POSIX', async () => {
    mockedOsPlatform.mockImplementationOnce(() => 'win32')
    await trianglePlugin.checkForTriangle()
    expect(mockedExeca).toHaveBeenCalledWith('where', ['triangle'])
    expect(mockedFsAccess.mock.calls[0][0]).toMatch(/\.exe$/)

    mockedOsPlatform.mockImplementationOnce(() => 'linux')
    await trianglePlugin.checkForTriangle()
    expect(mockedExeca).toHaveBeenCalledWith('type', ['triangle'])
  })

  test('bundled executable does not exist but triangle is globally installed', async () => {
    await expect(trianglePlugin.checkForTriangle()).resolves.toBeUndefined()
  })

  test('bundled executable does not exist, triangle not installed globally', async () => {
    mockedExeca.mockImplementationOnce(() => {
      throw new Error('not installed')
    })

    await expect(
      trianglePlugin.checkForTriangle()
    ).rejects.toThrowErrorMatchingSnapshot()
  })
})

describe('runTriangle', () => {
  // Source: https://www.mjt.me.uk/posts/smallest-png/
  const fileContent = Buffer.from(
    'iVBORw0KGgoAAAANSUhEUgAAAQAAAAEAAQMAAABmvDolAAAAA1BMVEW10NBjBBbqAAAAH0lEQVRoge3BAQ0AAADCoPdPbQ43oAAAAAAAAAAAvg0hAAABmmDh1QAAAABJRU5ErkJggg==',
    'base64'
  )

  afterEach(() => {
    mockedExeca.mockClear()
  })

  test('executes triangle with default config', async () => {
    const trianglePlugin = new TrianglePlugin({
      pluginOptions: {},
      options: {},
      sqipConfig: mockedConfig
    })
    await trianglePlugin.apply(fileContent, { ...mockedMetadata })
    expect(mockedExeca.mock.calls).toHaveLength(2)
    expect(mockedExeca.mock.calls[1]).toHaveLength(2)
    const args = mockedExeca.mock.calls[1][1] as string[] || []
    args[1] = "mocked"
    args[args.length -1] = "mocked"
    expect(args).toMatchSnapshot()
  })

  test('executes triangle with custom config, applying default number of triangles', async () => {
    const trianglePlugin = new TrianglePlugin({
      pluginOptions: { nf: 1 },
      options: {},

      sqipConfig: mockedConfig
    })
    await trianglePlugin.apply(fileContent, { ...mockedMetadata })
    expect(mockedExeca.mock.calls).toHaveLength(2)
    expect(mockedExeca.mock.calls[1]).toHaveLength(2)
    const args = mockedExeca.mock.calls[1][1] as string[] || []
    args[1] = "mocked"
    args[args.length -1] = "mocked"
    expect(args).toMatchSnapshot()
  })

  test('executes triangle with landscape dimensions', async () => {
    const trianglePlugin = new TrianglePlugin({
      pluginOptions: {},
      options: {},
      sqipConfig: mockedConfig
    })
    await trianglePlugin.apply(fileContent, {
      ...mockedMetadata,
      width: 600,
      height: 300
    })
    expect(mockedExeca.mock.calls).toHaveLength(2)
    expect(mockedExeca.mock.calls[1]).toHaveLength(2)
    const args = mockedExeca.mock.calls[1][1] as string[] || []
    args[1] = "mocked"
    args[args.length -1] = "mocked"
    expect(args).toMatchSnapshot()
  })

  test('allows hex as value for background', async () => {
    const trianglePlugin = new TrianglePlugin({
      pluginOptions: {
        bg: '#654321'
      },
      options: {},
      sqipConfig: mockedConfig
    })
    await trianglePlugin.apply(fileContent, { ...mockedMetadata })
    expect(mockedExeca.mock.calls).toHaveLength(2)
    expect(mockedExeca.mock.calls[1]).toHaveLength(2)
    const args = mockedExeca.mock.calls[1][1] as string[] || []
    args[1] = "mocked"
    args[args.length -1] = "mocked"
    expect(args).toMatchSnapshot()
  })
})
