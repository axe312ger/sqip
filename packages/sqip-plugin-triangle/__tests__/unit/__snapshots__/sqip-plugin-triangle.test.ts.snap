// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkForTriangle bundled executable does not exist, triangle not installed globally 1`] = `"Please ensure that triangle (https://github.com/esimov/triangle, written in Golang) is installed and globally available."`;

exports[`runTriangle allows hex as value for background 1`] = `
Array [
  "-in",
  "mocked",
  "-bl",
  "2",
  "-nf",
  "0",
  "-bf",
  "1",
  "-ef",
  "6",
  "-pr",
  "0.075",
  "-pth",
  "10",
  "-pts",
  "6",
  "-so",
  "10",
  "-wf",
  "0",
  "-st",
  "1",
  "-bg",
  "#654321",
  "-out",
  "mocked",
]
`;

exports[`runTriangle executes triangle with custom config, applying default number of triangles 1`] = `
Array [
  "-in",
  "mocked",
  "-bl",
  "2",
  "-nf",
  "1",
  "-bf",
  "1",
  "-ef",
  "6",
  "-pr",
  "0.075",
  "-pth",
  "10",
  "-pts",
  "6",
  "-so",
  "10",
  "-wf",
  "0",
  "-st",
  "1",
  "-bg",
  "#040204",
  "-out",
  "mocked",
]
`;

exports[`runTriangle executes triangle with default config 1`] = `
Array [
  "-in",
  "mocked",
  "-bl",
  "2",
  "-nf",
  "0",
  "-bf",
  "1",
  "-ef",
  "6",
  "-pr",
  "0.075",
  "-pth",
  "10",
  "-pts",
  "6",
  "-so",
  "10",
  "-wf",
  "0",
  "-st",
  "1",
  "-bg",
  "#040204",
  "-out",
  "mocked",
]
`;

exports[`runTriangle executes triangle with landscape dimensions 1`] = `
Array [
  "-in",
  "mocked",
  "-bl",
  "2",
  "-nf",
  "0",
  "-bf",
  "1",
  "-ef",
  "6",
  "-pr",
  "0.075",
  "-pth",
  "10",
  "-pts",
  "6",
  "-so",
  "10",
  "-wf",
  "0",
  "-st",
  "1",
  "-bg",
  "#040204",
  "-out",
  "mocked",
]
`;
